File: impl/typescript/src/main/ts/normalize.ts
export function normalize(input: string): string {
    return normalizeInternal(input, false);
}

export function normalizeGlob(input: string): string {
    return normalizeInternal(input, true);
}

function normalizeInternal(input: string, preserveGlobs: boolean): string {
    const output: string[] = [];
    let separatorPending = false;
    for (const character of input) {
        if (character === "'" || character === "’") {
            continue;
        }
        const codePoint = character.codePointAt(0)!;
        const alphanumeric = codePoint >= 0x61 && codePoint <= 0x7a
            || codePoint >= 0x41 && codePoint <= 0x5a
            || codePoint >= 0x30 && codePoint <= 0x39;
        const glob = preserveGlobs && (character === "*" || character === "?");
        if (alphanumeric || glob) {
            if (separatorPending && output.length > 0) {
                output.push(" ");
            }
            const lowered = codePoint >= 0x41 && codePoint <= 0x5a
                ? String.fromCodePoint(codePoint + 0x20)
                : character;
            if (lowered !== "*" || output.at(-1) !== "*") {
                output.push(lowered);
            }
            separatorPending = false;
        } else {
            separatorPending = true;
        }
    }
    return output.join("");
}

File: impl/lua/src/main/lua/liff/normalize.lua
local M = {}

local function normalize_internal(input, preserve_globs)
    local output = {}
    local separator_pending = false
    for _, code_point in utf8.codes(input) do
        if code_point ~= 0x27 and code_point ~= 0x2019 then
            local alphanumeric = code_point >= 0x61 and code_point <= 0x7a
                or code_point >= 0x41 and code_point <= 0x5a
                or code_point >= 0x30 and code_point <= 0x39
            local glob = preserve_globs and (code_point == 0x2a or code_point == 0x3f)
            if alphanumeric or glob then
                if separator_pending and #output > 0 then
                    output[#output + 1] = " "
                end
                if code_point >= 0x41 and code_point <= 0x5a then
                    code_point = code_point + 0x20
                end
                local character = string.char(code_point)
                if character ~= "*" or output[#output] ~= "*" then
                    output[#output + 1] = character
                end
                separator_pending = false
            else
                separator_pending = true
            end
        end
    end
    return table.concat(output)
end

function M.normalize(input)
    return normalize_internal(input, false)
end

function M.normalize_glob(input)
    return normalize_internal(input, true)
end

return M
